File: ConstructBTPreIn.java
import java.util.HashMap;

// Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

// Approach
//1) HashMap is used to store the inorder elements and its locations. Initialize index,start,end
//2) Make a root with preorder index and calculate its position in the map.
//3) Recursively call left by changing end and call right by changing start.

public class ConstructBTPreIn {
	public class TreeNode {
		int val;
		TreeNode left;
		TreeNode right;
		TreeNode() {}
		TreeNode(int val) { this.val = val; }
		TreeNode(int val, TreeNode left, TreeNode right) {
		    this.val = val;
		    this.left = left;
		    this.right = right;
		}
	}
	
	HashMap<Integer, Integer> map = new HashMap<>();
    int index;
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        for(int i = 0; i< inorder.length;i++){
            map.put(inorder[i],i);
        }
        return helper(preorder,inorder,0,preorder.length-1);
    }
    
    public TreeNode helper(int[] preorder, int[] inorder, int start, int end){
        if(start>end || index == preorder.length)return null;
        TreeNode root = new TreeNode(preorder[index]);
        int inIdx = map.get(root.val);
        index++;
        root.left = helper(preorder,inorder,start, inIdx-1);
        root.right = helper(preorder,inorder,inIdx+1,end);
        return root;
    }
}

File: ConstructBTPre_In.py
"""
Time Complexity : O(n)
Space Complexity : O(h) where h is the height of the tree
Did this code successfully run on Leetcode : Yes
"""
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        val_index_dict = { num:idx for idx, num in enumerate(inorder) }
        self.idx = 0
        def helper(start:int, end:int):
            # Base
            if start > end:
                return None
            # Logic
            root = TreeNode(preorder[self.idx])
            self.idx += 1
            rootIdx = val_index_dict[root.val]
            root.left = helper(start, rootIdx - 1)
            root.right = helper(rootIdx + 1, end)
            return root
        return helper(0, len(inorder) - 1)
        